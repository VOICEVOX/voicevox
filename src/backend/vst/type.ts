import { TrackId } from "@/type/preload";

export type Routing = {
  channelMode: "mono" | "stereo";
  channelIndex: Record<TrackId, number>;
};
