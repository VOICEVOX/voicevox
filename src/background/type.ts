import { z } from "zod";
import { engineIdSchema } from "@/type/preload";

// デフォルトエンジン情報
export const defaultEngineInfosEnvSchema = z
  .object({
    uuid: engineIdSchema,
    host: z.string(),
    name: z.string(),
    executionEnabled: z.boolean(),
    executionFilePath: z.string(),
    executionArgs: z.array(z.string()),
    path: z.string().optional(),
  })
  .array();
