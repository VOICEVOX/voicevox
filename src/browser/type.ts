import { IpcIHData } from "@/type/ipc";

export type WorkerToMainMessage = {
  [K in keyof IpcIHData]: {
    type: K;
    return: IpcIHData[K]["return"];
    eventId: string;
  };
}[keyof IpcIHData];

export type MainToWorkerMessage = {
  [K in keyof IpcIHData]: {
    type: K;
    args: IpcIHData[K]["args"];
    eventId: string;
  };
}[keyof IpcIHData];
