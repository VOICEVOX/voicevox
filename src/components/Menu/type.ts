export type MenuItemBase<T extends string> = {
  type: T;
  label?: string;
};

export type MenuItemSeparator = MenuItemBase<"separator">;

export type MenuItemRoot = MenuItemBase<"root"> & {
  onClick?: () => void;
  subMenu: MenuItemData[];
  icon?: string;
  disabled?: boolean;
  disableWhenUiLocked: boolean;
  disablreloadingLocked?: boolean;
};

export type MenuItemButton = MenuItemBase<"button"> & {
  onClick: () => void;
  icon?: string;
  disabled?: boolean;
  disableWhenUiLocked: boolean;
  disablreloadingLocked?: boolean;
};

export type MenuItemData = MenuItemSeparator | MenuItemRoot | MenuItemButton;

export type MenuItemType = MenuItemData["type"];
