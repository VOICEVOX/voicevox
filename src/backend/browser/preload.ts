import { SandboxKey, Sandbox } from "../type/preload";
import { api } from "./sandbox";

const sandbox: Sandbox = api;
// @ts-expect-error readonlyになっているが、初期化処理はここで行うので問題ない
window[SandboxKey] = sandbox;
