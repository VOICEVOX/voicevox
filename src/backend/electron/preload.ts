import { contextBridge, ipcRenderer } from "electron";
import type { IpcRendererInvoke } from "./ipc";
import {
  ConfigType,
  EngineId,
  Sandbox,
  SandboxKey,
  TextAsset,
} from "@/type/preload";

const ipcRendererInvokeProxy = new Proxy(
  {},
  {
    get:
      (_, channel: string) =>
      (...args: unknown[]) =>
        ipcRenderer.invoke(channel, ...args),
  },
) as IpcRendererInvoke;

const api: Sandbox = {
  getAppInfos: async () => {
    return await ipcRendererInvokeProxy.GET_APP_INFOS();
  },

  getTextAsset: (textType) => {
    return ipcRendererInvokeProxy.GET_TEXT_ASSET(textType) as Promise<
      TextAsset[typeof textType]
    >;
  },

  getAltPortInfos: async () => {
    return await ipcRendererInvokeProxy.GET_ALT_PORT_INFOS();
  },

  showAudioSaveDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_AUDIO_SAVE_DIALOG({
      title,
      defaultPath,
    });
  },

  showTextSaveDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_TEXT_SAVE_DIALOG({ title, defaultPath });
  },

  showSaveDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_SAVE_DIRECTORY_DIALOG({ title });
  },

  showVvppOpenDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_VVPP_OPEN_DIALOG({ title, defaultPath });
  },

  showOpenDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_OPEN_DIRECTORY_DIALOG({ title });
  },

  showProjectSaveDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_PROJECT_SAVE_DIALOG({
      title,
      defaultPath,
    });
  },

  showProjectLoadDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_PROJECT_LOAD_DIALOG({ title });
  },

  showImportFileDialog: ({ title, name, extensions }) => {
    return ipcRendererInvokeProxy.SHOW_IMPORT_FILE_DIALOG({
      title,
      name,
      extensions,
    });
  },

  writeFile: async ({ filePath, buffer }) => {
    return await ipcRendererInvokeProxy.WRITE_FILE({ filePath, buffer });
  },

  readFile: async ({ filePath }) => {
    return await ipcRendererInvokeProxy.READ_FILE({ filePath });
  },

  isAvailableGPUMode: () => {
    return ipcRendererInvokeProxy.IS_AVAILABLE_GPU_MODE();
  },

  isMaximizedWindow: () => {
    return ipcRendererInvokeProxy.IS_MAXIMIZED_WINDOW();
  },

  onReceivedIPCMsg: (listeners) => {
    Object.entries(listeners).forEach(([channel, listener]) => {
      ipcRenderer.on(channel, listener);
    });
  },

  closeWindow: () => {
    void ipcRendererInvokeProxy.CLOSE_WINDOW();
  },

  minimizeWindow: () => {
    void ipcRendererInvokeProxy.MINIMIZE_WINDOW();
  },

  maximizeWindow: () => {
    void ipcRendererInvokeProxy.MAXIMIZE_WINDOW();
  },

  logError: (...params) => {
    console.error(...params);
    // 経緯 https://github.com/VOICEVOX/voicevox/pull/1620#discussion_r1371804569
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "error",
    });
  },

  logWarn: (...params) => {
    console.warn(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "warn",
    });
  },

  logInfo: (...params) => {
    console.info(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "info",
    });
  },

  openLogDirectory: () => {
    void ipcRendererInvokeProxy.OPEN_LOG_DIRECTORY();
  },

  engineInfos: () => {
    return ipcRendererInvokeProxy.ENGINE_INFOS();
  },

  restartEngine: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.RESTART_ENGINE({ engineId });
  },

  openEngineDirectory: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.OPEN_ENGINE_DIRECTORY({ engineId });
  },

  checkFileExists: (file) => {
    return ipcRendererInvokeProxy.CHECK_FILE_EXISTS({ file });
  },

  changePinWindow: () => {
    void ipcRendererInvokeProxy.CHANGE_PIN_WINDOW();
  },

  hotkeySettings: (newData) => {
    return ipcRendererInvokeProxy.HOTKEY_SETTINGS({ newData });
  },

  getDefaultHotkeySettings: async () => {
    return await ipcRendererInvokeProxy.GET_DEFAULT_HOTKEY_SETTINGS();
  },

  getDefaultToolbarSetting: async () => {
    return await ipcRendererInvokeProxy.GET_DEFAULT_TOOLBAR_SETTING();
  },

  setNativeTheme: (source) => {
    void ipcRendererInvokeProxy.SET_NATIVE_THEME(source);
  },

  vuexReady: () => {
    void ipcRendererInvokeProxy.ON_VUEX_READY();
  },

  /**
   * 設定情報を取得する
   */
  getSetting: async (key) => {
    return (await ipcRendererInvokeProxy.GET_SETTING(
      key,
    )) as ConfigType[typeof key];
  },

  /**
   * 設定情報を保存する
   */
  setSetting: async (key, newValue) => {
    return (await ipcRendererInvokeProxy.SET_SETTING(
      key,
      newValue,
    )) as typeof newValue;
  },

  setEngineSetting: async (engineId, engineSetting) => {
    await ipcRendererInvokeProxy.SET_ENGINE_SETTING(engineId, engineSetting);
  },

  installVvppEngine: async (filePath) => {
    return await ipcRendererInvokeProxy.INSTALL_VVPP_ENGINE(filePath);
  },

  uninstallVvppEngine: async (engineId) => {
    return await ipcRendererInvokeProxy.UNINSTALL_VVPP_ENGINE(engineId);
  },

  validateEngineDir: async (engineDir) => {
    return await ipcRendererInvokeProxy.VALIDATE_ENGINE_DIR({ engineDir });
  },

  /**
   * アプリを再読み込みする。
   * 画面以外の情報を刷新する。
   */
  reloadApp: async ({ isMultiEngineOffMode }) => {
    await ipcRendererInvokeProxy.RELOAD_APP({ isMultiEngineOffMode });
  },
};

contextBridge.exposeInMainWorld(SandboxKey, api);
