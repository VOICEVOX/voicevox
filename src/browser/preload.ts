import { SandboxKey } from "../type/preload";
import { api } from "./sandbox";

// @ts-expect-error readonlyになっているが、初期化処理はここで行うので問題ない
window[SandboxKey] = api;
