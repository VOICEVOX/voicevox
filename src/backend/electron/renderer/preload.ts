import { contextBridge, ipcRenderer, webUtils } from "electron";
import {
  wrapToTransferableResult,
  getOrThrowTransferableResult,
} from "../transferableResultHelper";
import { type IpcRendererInvoke } from "./ipc";
import { BridgeKey, SandboxWithTransferableResult } from "./backendApiLoader";
import { ConfigType, EngineId, Sandbox, TextAsset } from "@/type/preload";

const ipcRendererInvokeProxy = new Proxy(
  {},
  {
    get:
      (_, channel: string) =>
      async (...args: unknown[]) => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        const transferableResult = await ipcRenderer.invoke(channel, ...args);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        return getOrThrowTransferableResult(transferableResult);
      },
  },
) as IpcRendererInvoke;

const api: Sandbox = {
  getTextAsset: (textType) => {
    return ipcRendererInvokeProxy.GET_TEXT_ASSET(textType) as Promise<
      TextAsset[typeof textType]
    >;
  },

  getAltPortInfos: async () => {
    return await ipcRendererInvokeProxy.GET_ALT_PORT_INFOS();
  },

  getInitialProjectFilePath: async () => {
    return await ipcRendererInvokeProxy.GET_INITIAL_PROJECT_FILE_PATH();
  },

  showSaveDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_SAVE_DIRECTORY_DIALOG({ title });
  },

  showOpenDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_OPEN_DIRECTORY_DIALOG({ title });
  },

  showOpenFileDialog: ({ title, name, extensions, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_OPEN_FILE_DIALOG({
      title,
      name,
      extensions,
      defaultPath,
    });
  },

  showSaveFileDialog: ({ title, name, extensions, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_SAVE_FILE_DIALOG({
      title,
      name,
      extensions,
      defaultPath,
    });
  },

  writeFile: async ({ filePath, buffer }) => {
    return await ipcRendererInvokeProxy.WRITE_FILE({ filePath, buffer });
  },

  readFile: async ({ filePath }) => {
    return await ipcRendererInvokeProxy.READ_FILE({ filePath });
  },

  isAvailableGPUMode: () => {
    return ipcRendererInvokeProxy.IS_AVAILABLE_GPU_MODE();
  },

  isMaximizedWindow: () => {
    return ipcRendererInvokeProxy.IS_MAXIMIZED_WINDOW();
  },

  onReceivedIPCMsg: (listeners) => {
    Object.entries(listeners).forEach(([channel, listener]) => {
      ipcRenderer.on(channel, listener);
    });
  },

  closeWindow: () => {
    void ipcRendererInvokeProxy.CLOSE_WINDOW();
  },

  minimizeWindow: () => {
    void ipcRendererInvokeProxy.MINIMIZE_WINDOW();
  },

  toggleMaximizeWindow: () => {
    void ipcRendererInvokeProxy.TOGGLE_MAXIMIZE_WINDOW();
  },

  toggleFullScreen: () => {
    void ipcRendererInvokeProxy.TOGGLE_FULLSCREEN();
  },

  zoomIn: () => {
    void ipcRendererInvokeProxy.ZOOM_IN();
  },
  zoomOut: () => {
    void ipcRendererInvokeProxy.ZOOM_OUT();
  },
  zoomReset: () => {
    void ipcRendererInvokeProxy.ZOOM_RESET();
  },

  logError: (...params) => {
    console.error(...params);
    // 経緯 https://github.com/VOICEVOX/voicevox/pull/1620#discussion_r1371804569
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "error",
    });
  },

  logWarn: (...params) => {
    console.warn(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "warn",
    });
  },

  logInfo: (...params) => {
    console.info(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "info",
    });
  },

  openLogDirectory: () => {
    void ipcRendererInvokeProxy.OPEN_LOG_DIRECTORY();
  },

  engineInfos: () => {
    return ipcRendererInvokeProxy.ENGINE_INFOS();
  },

  restartEngine: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.RESTART_ENGINE({ engineId });
  },

  openEngineDirectory: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.OPEN_ENGINE_DIRECTORY({ engineId });
  },

  checkFileExists: (file) => {
    return ipcRendererInvokeProxy.CHECK_FILE_EXISTS({ file });
  },

  changePinWindow: () => {
    void ipcRendererInvokeProxy.CHANGE_PIN_WINDOW();
  },

  hotkeySettings: (newData) => {
    return ipcRendererInvokeProxy.HOTKEY_SETTINGS({ newData });
  },

  getDefaultToolbarSetting: async () => {
    return await ipcRendererInvokeProxy.GET_DEFAULT_TOOLBAR_SETTING();
  },

  setNativeTheme: (source) => {
    void ipcRendererInvokeProxy.SET_NATIVE_THEME(source);
  },

  vuexReady: () => {
    void ipcRendererInvokeProxy.ON_VUEX_READY();
  },

  /**
   * 設定情報を取得する
   */
  getSetting: async (key) => {
    return (await ipcRendererInvokeProxy.GET_SETTING(
      key,
    )) as ConfigType[typeof key];
  },

  /**
   * 設定情報を保存する
   */
  setSetting: async (key, newValue) => {
    return (await ipcRendererInvokeProxy.SET_SETTING(
      key,
      newValue,
    )) as typeof newValue;
  },

  setEngineSetting: async (engineId, engineSetting) => {
    await ipcRendererInvokeProxy.SET_ENGINE_SETTING(engineId, engineSetting);
  },

  installVvppEngine: async (filePath) => {
    return await ipcRendererInvokeProxy.INSTALL_VVPP_ENGINE(filePath);
  },

  uninstallVvppEngine: async (engineId) => {
    return await ipcRendererInvokeProxy.UNINSTALL_VVPP_ENGINE(engineId);
  },

  validateEngineDir: async (engineDir) => {
    return await ipcRendererInvokeProxy.VALIDATE_ENGINE_DIR({ engineDir });
  },

  /**
   * アプリを再読み込みする。
   * 画面以外の情報を刷新する。
   */
  reloadApp: async ({ isMultiEngineOffMode }) => {
    await ipcRendererInvokeProxy.RELOAD_APP({ isMultiEngineOffMode });
  },

  /** webUtils.getPathForFileを呼ぶ */
  getPathForFile: (file) => {
    return webUtils.getPathForFile(file);
  },
};

const wrapApi = (baseApi: Sandbox): SandboxWithTransferableResult => {
  const wrappedApi = {} as SandboxWithTransferableResult;
  for (const key in baseApi) {
    const propKey = key as keyof SandboxWithTransferableResult;
    // @ts-expect-error とりあえず動くので無視
    wrappedApi[propKey] = async (...args: unknown[]) => {
      // @ts-expect-error とりあえず動くので無視
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      return wrapToTransferableResult(() => baseApi[propKey](...args));
    };
  }
  return wrappedApi;
};
contextBridge.exposeInMainWorld(BridgeKey, wrapApi(api));
