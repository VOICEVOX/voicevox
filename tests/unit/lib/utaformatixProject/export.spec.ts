import { it, expect } from "vitest";
import { ufProjectFromVoicevox } from "@/sing/utaformatixProject/fromVoicevox";
import {
  createDefaultTempo,
  createDefaultTimeSignature,
  createDefaultTrack,
} from "@/sing/domain";
import { NoteId } from "@/type/preload";

const createNoteId = () => NoteId(crypto.randomUUID());

it("トラックを変換できる", async () => {
  const track = createDefaultTrack();
  track.notes.push({
    id: createNoteId(),
    noteNumber: 60,
    position: 0,
    duration: 480,
    lyric: "ど",
  });

  const project = ufProjectFromVoicevox(
    {
      tracks: [track],
      tpqn: 480,
      tempos: [createDefaultTempo(0)],
      timeSignatures: [createDefaultTimeSignature(1)],
    },
    "test",
  );

  const ufData = project.toUfDataObject();

  expect(ufData).toMatchSnapshot();
});
