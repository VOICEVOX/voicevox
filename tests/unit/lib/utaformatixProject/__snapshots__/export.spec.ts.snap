// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`トラックを変換できる 1`] = `
{
  "formatVersion": 1,
  "project": {
    "measurePrefix": 0,
    "name": "test",
    "tempos": [
      {
        "bpm": 120,
        "tickPosition": 0,
      },
    ],
    "timeSignatures": [
      {
        "denominator": 4,
        "measurePosition": 1,
        "numerator": 4,
      },
    ],
    "tracks": [
      {
        "name": "無名トラック",
        "notes": [
          {
            "key": 60,
            "lyric": "ど",
            "tickOff": 480,
            "tickOn": 0,
          },
        ],
      },
    ],
  },
}
`;
